### PR TITLE
CMake3 port and various cleanups

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

**Comment by tribal-tec** [comment]
FYI @rdumusc 
Good luck!

File: Common.cmake [hunk 1]
```diff
@@ -1,90 +1,84 @@
...
+# - PATCH releases preseve forward+backward API compatibility (no new features)
+if(NOT DEFINED ${PROJECT_NAME}_VERSION_ABI)
+  set(${PROJECT_NAME}_VERSION_ABI ${${PROJECT_NAME}_VERSION_MAJOR})
+  message(STATUS "VERSION_ABI not set for ${PROJECT_NAME}. Using VERSION_MAJOR=${${PROJECT_NAME}_VERSION_MAJOR} as the ABI version.")
+endif()
 
 if(CMAKE_SYSTEM_NAME MATCHES "Linux")
   include(LSBInfo)
   set(LINUX TRUE)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)  # -fPIC
```
Comment (eile) [inline Common.cmake]: can be set outside of if(linux)?

Comment (tribal-tec) [inline Common.cmake]: probably. Was like this before

File: CommonCompiler.cmake [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  set(COMMON_CXX_FLAGS
+    -Wnon-virtual-dtor -Wsign-promo -Wvla -fno-strict-aliasing)
+
+  if(NOT WIN32 AND NOT XCODE_VERSION)
+    list(APPEND COMMON_C_FLAGS -Werror)
+  endif()
+
+  if(GCC_COMPILER_VERSION VERSION_GREATER 4.1)
+    list(APPEND COMMON_C_FLAGS -Wshadow)
+  endif()
```
Comment (eile) [inline CommonCompiler.cmake]: rm if?

File: CommonCompiler.cmake [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    if(CMAKE_VERSION VERSION_LESS 3.3)
+      target_compile_options(${Name} PRIVATE
+        "$<$<CONFIG:Debug>:${COMMON_CXX_FLAGS_DEBUG}>"
+        "$<$<CONFIG:RelWithDebInfo>:${COMMON_CXX_FLAGS_RELWITHDEBINFO}>"
+        "$<$<CONFIG:Release>:${COMMON_CXX_FLAGS_RELEASE}>"
+        "${COMMON_CXX_FLAGS}"
+        "$<$<CONFIG:Debug>:${COMMON_C_FLAGS_DEBUG}>"
+        "$<$<CONFIG:RelWithDebInfo>:${COMMON_C_FLAGS_RELWITHDEBINFO}>"
+        "$<$<CONFIG:Release>:${COMMON_C_FLAGS_RELEASE}>"
+        "${COMMON_C_FLAGS}"
```
Comment (eile) [inline CommonCompiler.cmake]: (puke)

File: CommonDocumentation.cmake [hunk 4]
```diff
@@ -1,60 +0,0 @@
...
-        CONFIGURATIONS Release)
-    endforeach()
-    # need at least one file for 'make install'
-    install(FILES index.html DESTINATION share/${PROJECT_NAME}
-      CONFIGURATIONS Release)
-  else()
-    # Buildyard expects an install target for all projects
-    install(CODE "MESSAGE(\"Nothing to install, done.\")")
-  endif()
-endfunction()
```
Comment (eile) [inline CommonDocumentation.cmake]: Afair this is still used for BBPDocumentation.

**Comment by eile** [comment]
Good lord -- quite a few gotchas to not miss in various projects.

**Comment by eile** [comment]
+1 otherwise from what I could grasp

File: CommonDocumentation.cmake (hunk 4 above)
Comment (tribal-tec) [inline CommonDocumentation.cmake]: Yes, I moved the file there.

**Comment by tribal-tec** [comment]
I built with gcc, clang and icc with various different options on Ubuntu. Let's see what breaks on RHEL and OSX :)

**Pull request merged**